### DISCUSSION TITLE
20220526

### DISCUSSION THREAD

**Discussion by tmiw** [ARSFI/paclink-open-source#25] [General]
This is a pre-production build intended to gauge current status of the project and determine what needs to be fixed. Brief testing indicates that packet (using Kenwood TM-D700A), AGW (using AGWPE and UZ7HO soundmodem) and Telnet channels work; Pactor is TBD.

Contains (vs. [last pre-release version](https://github.com/ARSFI/paclink-open-source/releases/tag/nightly_20220525)):

* Help file now properly displays when choosing Help->Show Contents.
* Incoming emails with multiple addresses in the To field are now properly parsed and stored.

Instructions:

1. Unzip paclink-open-source.zip somewhere on your hard drive.
2. Run setup.exe to install on your system. Paclink will be installed in the Start menu.
3. Open Paclink from the Start menu and reconfigure if required. (There is no import functionality from the existing Paclink at this time but if you used a pre-release version, configuration should be saved.)

Please leave comments in the release's discussion thread or email K6AQ@winlink.org with feedback. 👍
 

<hr /><em>This discussion was created from the release <a href='https://github.com/ARSFI/paclink-open-source/releases/tag/nightly_20220526'>20220526</a>.</em>

**Comment by kd4vrf** [comment]
I tested the "help --> contents..." and it worked as expected for me.

I was also able to test an email with three recipients in the "to:" field and it worked as expected. I will check it again next week with the 400+ message from the Winlink Wednesday checkin and let you know how it goes.

I did have another instance of the "Paclink has stopped working" crash. It occurred at just shy of four hours of up time. See the screen shot below:

![Screen Shot 2022-05-27 at 12 39 53 PM](https://user-images.githubusercontent.com/45251519/170792448-dab76302-b732-47fa-ada7-d8daa3a62b77.png)

I had to reboot from that crash to recover, it would not connect to the Pactor modem until I rebooted windows. I tried rebooting just the modem but that didn't resolve it.

Here are the logs from that crash:
[Logs.zip](https://github.com/ARSFI/paclink-open-source/files/8789880/Logs.zip)

Let me know if you need anything else to try to debug the crash.

**Reply by tmiw** [reply]
Unfortunately I'm not seeing anything obvious in the logs. Can you check the Windows Event Viewer and see if there's anything Paclink related in there?

Additionally, does this crash happen if you don't use Pactor? If not, it would at least narrow possible causes down.

**Reply by kd4vrf** [reply]
I've only had significant time testing with Pactor at this point, I only ran telnet during my initial tests and that was only for a couple of tests. I didn't see this when I was running telnet, admittedly there wasn't much of an opportunity to see the error.

I did see this in the event viewer at the time of the error:

```
Application: Paclink.exe
CoreCLR Version: 6.0.522.21309
.NET Version: 6.0.5
Description: The process was terminated due to an unhandled exception.
Exception Info: System.NullReferenceException: Object reference not set to an instance of an object.
   at Paclink.SCSHostPort.SendNextFrame(Boolean blnRepeatLast) in Z:\mooneer\paclink-open-source\Paclink\ModemSCS.cs:line 2885
   at Paclink.SCSHostPort.PollOutgoing() in Z:\mooneer\paclink-open-source\Paclink\ModemSCS.cs:line 2246
   at Paclink.PrimaryThread.ChannelThread() in Z:\mooneer\paclink-open-source\Paclink\PrimaryThread.cs:line 278
   at System.Threading.Thread.StartCallback()
```

It looks like the issue is in ModemSCS.cs and that looks like it handles all of the Pactor modems. So it looks to be Pactor related.

Attached are the exported logs from event viewer at the time of the error and my attempt to recover by rebooting the Pactor modem only, and then a Windows VM reboot.
[Packlink_has_stopped_working.zip](https://github.com/ARSFI/paclink-open-source/files/8805747/Packlink_has_stopped_working.zip)

**Reply by tmiw** [reply]
Hmm, based on what I'm seeing in SendNextFrame, it seems like a threading issue. There are a few outbound queues that aren't wrapped with `Queue.Synchronized()`, so there could be issues if one part of the application reads from them while another also updates them. I wrapped those remaining queues and put out another release at https://github.com/ARSFI/paclink-open-source/releases/tag/nightly_20220531; hopefully that goes a bit longer before crashing at least.
